src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -0,0 +1,702 @@
+                    goto map_fail;
+
+                mpi_errno =
+                    MPIR_Bcast(serialized_hnd, MPL_SHM_GHND_SZ, MPI_BYTE, 0,
+                               shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                /* ensure all other processes have mapped successfully */
+                mpi_errno = MPIR_Allreduce(map_result_ptr, &all_map_result, 1, MPI_INT,
+                                           MPI_MAX, shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                if (all_map_result != SYMSHM_SUCCESS)
+                    goto map_fail;
+            } else {
+                char serialized_hnd[MPL_SHM_GHND_SZ] = { 0 };
+
+                mpi_errno = MPIR_Bcast(map_result_ptr, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                if (*map_result_ptr != SYMSHM_SUCCESS)
+                    goto map_fail;
+
+                mpi_errno =
+                    MPIR_Bcast(serialized_hnd, MPL_SHM_GHND_SZ, MPI_BYTE, 0,
+                               shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                mpl_err =
+                    MPL_shm_hnd_deserialize(shm_seg->hnd, serialized_hnd, strlen(serialized_hnd));
+                if (mpl_err) {
+                    *map_result_ptr = SYMSHM_OTHER_FAIL;
+                    goto result_sync;
+                }
+
+                mpl_err = MPL_shm_fixed_seg_attach(shm_seg->hnd, shm_seg->segment_len,
+                                                   (void **) &(shm_seg->base_addr), 0);
+                if (mpl_err) {
+                    *map_result_ptr =
+                        (mpl_err == MPL_SHM_EINVAL) ? SYMSHM_MAP_FAIL : SYMSHM_OTHER_FAIL;
+                } else
+                    mapped_flag = true;
+
+              result_sync:
+                /* check results of all processes. If any failure happens (max result > 0),
+                 * return SYMSHM_OTHER_FAIL if anyone reports it (max result == 2).
+                 * Otherwise return SYMSHM_MAP_FAIL (max result == 1). */
+                mpi_errno = MPIR_Allreduce(map_result_ptr, &all_map_result, 1, MPI_INT,
+                                           MPI_MAX, shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                if (all_map_result != SYMSHM_SUCCESS)
+                    goto map_fail;
+            }
+
+            mpi_errno = MPIR_Barrier(shm_comm_ptr, &errflag);
+            MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+            if (shm_comm_ptr->rank == 0) {
+                /* shm_seg->hnd no longer needed */
+                mpl_err = MPL_shm_seg_remove(shm_seg->hnd);
```

Comment:
I think `/* unlink shared memory region so it gets deleted when all processes exit */` is more specific than `no longer needed`. Can you please use the original comment?

Besides, can we move `MPIR_Allreduce` out from the if-else branch? It looks like the code is the same for both root and others. 

I do not think `MPIR_Barrier` is needed here. `Allreduce` implicitly ensures that all processes have made the call.

Reply:
done